### ISSUE TITLE
entity title system

### ISSUE THREAD

**Issue by end-user** [Zergatul/cheatutils#44]
I've been trying to use the entity title system without much success.  Specifically, I was trying to show the owners, but I really can't get anything to work.  I chose the Tamable class (currently cats, dogs, parrots). I enabled all the options, but the only thing that seems to work is glow.  I'm not seeing any errors.

**Comment by Zergatul** [comment]
what is your version? in older versions you have to enable checkbox in entity title module, but later it was removed and now each entry in entity ESP has its own checkbox names custom title system
maybe conflict with other mod? did you try it in vanilla?

**Comment by end-user** [comment]
v2.7.2 and yeah, it looks like Sodium blocks all the features except glow.

**Comment by Zergatul** [comment]
Interesting. Do other mods with similar drawing capabilities work? Like baritone

**Comment by end-user** [comment]
I've never used baritone to draw/didn't know it could.  I have used jsmacros to do 3D drawing though.  And flight HUD I think uses 2D.  Part of the reason I was looking at this is I use 'pet owner' which does the same thing. It's nice when I can reduce redundant mods.

**Comment by Zergatul** [comment]
The problem is these optimization mods are developed by some OpenGL graphics geniuses, but I barely understand how code for drawing single works 😁

**Comment by end-user** [comment]
Totally get it.  My understanding of what jsmacros is doing is basically providing hooks for (and parsing of) other languages to make native MC code calls.  It allows novices like me to write (for example) simple javascript that can perform in-game functions.  For example, I have one that allows me to highlight specific blocks.

Since it works with sodium, there must be something it's doing that you're not.  Is it calling an interface rather than a class instance, for example?  If sodium completely replaces the class and you're making a direct call to something that's been overwritten, maybe? I'm just grasping.

**Comment by Zergatul** [comment]
Drawing block outline must be done via calling standard minecraft method.
Drawing lines for ESP done via using lower level code, because such 1px 3d lines are not used in minecraft client.

**Comment by end-user** [comment]
Yeah, that's probably true for the tracers, but the outline can be 3D.

**Comment by end-user** [comment]
For example, in jsmacros the [box rendering](https://github.com/JsMacros/JsMacros/blob/main/src/main/java/xyz/wagyourtail/jsmacros/client/api/classes/render/components3d/Box.java) seems similar, though it looks like each box is responsible for rendering itself.  It looks like there's a call to the Tessellator draw() function; I don't know if that's needed. The project appears that there is code for creating tracers too.  Don't know if it's useful to compare approaches.

**Comment by Zergatul** [comment]
> Yeah, that's probably true for the tracers, but the outline can be 3D.

All ESP lines are drawn in 3d context. Entity titles are drawn in 2d context. Entity glow reuses existing minecraft functionality.

> The project appears that there is code for creating tracers too. Don't know if it's useful to compare approaches.

jsmacros work for you with Sodium? Can you show example what does it draw? I never used this mod.

**Comment by end-user** [comment]
Yeah, jsmacros works with Sodium.  Here's an example where I drew boxes around double chests.  I'm doing it for each block rather than trying to figure out how to connect them.  You can also see that you can turn off culling so the boxes show through.
![image](https://github.com/Zergatul/cheatutils/assets/4923193/90e5a52a-2bdb-4f57-a13c-9b1269946301)

**Comment by Zergatul** [comment]
I tried Sodium by myself, and nothing from cheatutils is rendered. Maybe I am doing rendering calls in the wrong time. But I checked baritone, and their ESP-like rendering works with Sodium. Their line drawing code is not very different from mine, but there are some differences.
Sorry, this stuff is pretty boring for me, and I don't play minecraft now. I am not sure when I will be able to work on this.
